File: Day 022/Problem 1110. Delete Nodes And Return Forest.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<TreeNode*> delNodes(TreeNode* root, vector<int>& to_delete) {
        vector<TreeNode*> result;
        map<int, bool> shouldDelete;

        for (int node : to_delete) {
            shouldDelete[node] = true;
        }

        helper(root, shouldDelete, result, false);

        return result;
    }

    TreeNode* helper(TreeNode* root, map<int, bool> shouldDelete, vector<TreeNode*> &result, bool included) {
        if (!root) return nullptr;

        if (shouldDelete[root->val] == false) {
            if (!included) result.push_back(root);
            root->left = helper(root->left, shouldDelete, result, true);
            root->right = helper(root->right, shouldDelete, result, true);
            return root;
        }

        helper(root->left, shouldDelete, result, false);
        helper(root->right, shouldDelete, result, false);

        return nullptr;
    }
};

File: Day 095/Problem 1110. Delete Nodes And Return Forest.java
import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode() {}
    TreeNode(int val) { this.val = val; }
    TreeNode(int val, TreeNode left, TreeNode right) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

class Solution {
    public List<TreeNode> delNodes(TreeNode root, int[] to_delete) {
        Set<Integer> toDeleteSet = new HashSet<>();
        for (int node : to_delete) toDeleteSet.add(node);

        List<TreeNode> forest = new ArrayList<>();
        root = dfs(root, toDeleteSet, forest);

        if (root != null) {
            forest.add(root);
        }

        return forest;
    }

    private TreeNode dfs(TreeNode root, Set<Integer> toDeleteSet, List<TreeNode> forest) {
        if (root == null) return null;
        
        root.left = dfs(root.left, toDeleteSet, forest);
        root.right = dfs(root.right, toDeleteSet, forest);

        if (toDeleteSet.contains(root.val)) {
            if (root.left != null) forest.add(root.left);
            if (root.right != null) forest.add(root.right);

            return null;
        }

        return root;
    }
}